pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/strategy/ReplicaGroupSegmentAssignmentStrategy.java

```diff
@@ -0,0 +1,135 @@
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.TreeMap;
+import org.apache.helix.HelixManager;
+import org.apache.pinot.common.assignment.InstancePartitions;
+import org.apache.pinot.controller.helix.core.assignment.segment.SegmentAssignmentUtils;
+import org.apache.pinot.spi.config.table.ReplicaGroupStrategyConfig;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.config.table.TableType;
+import org.apache.pinot.spi.config.table.assignment.InstancePartitionsType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+class ReplicaGroupSegmentAssignmentStrategy implements SegmentAssignmentStrategy {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ReplicaGroupSegmentAssignmentStrategy.class);
+
+  private static HelixManager _helixManager;
+  private static String _tableName;
+  private static String _partitionColumn;
+  private int _replication;
+  private TableConfig _tableConfig;
+
+  @Override
+  public void init(HelixManager helixManager, TableConfig tableConfig) {
+    _helixManager = helixManager;
+    _tableConfig = tableConfig;
+    _tableName = tableConfig.getTableName();
+    _replication = tableConfig.getValidationConfig().getReplicationNumber();
+    ReplicaGroupStrategyConfig replicaGroupStrategyConfig =
+        tableConfig.getValidationConfig().getReplicaGroupStrategyConfig();
+    _partitionColumn = replicaGroupStrategyConfig != null ? replicaGroupStrategyConfig.getPartitionColumn() : null;
+    if (_partitionColumn == null) {
+      LOGGER.info("Initialized ReplicaGroupSegmentAssignmentStrategy "
+              + "with replication: {} without partition column for table: {} ", _replication, _tableName);
+    } else {
+      LOGGER.info("Initialized ReplicaGroupSegmentAssignmentStrategy "
+              + "with replication: {} and partition column: {} for table: {}",
+              _replication, _partitionColumn, _tableName);
+    }
+  }
+
+  /**
+   * Assigns the segment for the replica-group based segment assignment strategy and returns the assigned instances.
+   */
+  @Override
+  public List<String> assignSegment(String segmentName, Map<String, Map<String, String>> currentAssignment,
+      InstancePartitions instancePartitions, InstancePartitionsType instancePartitionsType) {
+    int numPartitions = instancePartitions.getNumPartitions();
+    SegmentAssignmentStrategyUtils.checkReplication(instancePartitions, _replication, _tableName);
+    int partitionId;
+    if (_partitionColumn == null || numPartitions == 1) {
+      partitionId = 0;
+    } else {
+      // Uniformly spray the segment partitions over the instance partitions
+      if (_tableConfig.getTableType() == TableType.OFFLINE) {
+        partitionId =
+            SegmentAssignmentUtils
+                .getOfflineSegmentPartitionId(segmentName, _tableName, _helixManager, _partitionColumn) % numPartitions;
```

Comment:
Same here, we have a bunch of partition functions to generate the partition Ids already. We should consider reusing them if possible.

Reply:
partitionId function is calculated different for Realtime and Offline Segment Assignments. I tried refactoring but it seems more complicated to merge